### DISCUSSION TITLE
How to know the number of tcp connections created based on the configuration options ?

### DISCUSSION THREAD

**Discussion by kkattamudi-ms** [Azure/azure-cosmos-dotnet-v3#4741] [Q&A]
Even we restricted the cosmos db client objects to be 6 and with maximum TCP Connections to be created as 100 . We are seeing huge number of TCP Connections are getting created due to which the performace of our worker is getting degraded .
Please find the code sample we are using :
**CosmosDB Client Options :**
        "MaxRequestsPerTcpConnection": 10,
        "MaxTcpConnectionsPerEndpoint": 100

            this.cosmosClient = await cosmosClientBuilder
                .WithConnectionModeDirect(maxRequestsPerTcpConnection: this.options.ConnectionPolicy.MaxRequestsPerTcpConnection, maxTcpConnectionsPerEndpoint: this.options.ConnectionPolicy.MaxTcpConnectionsPerEndpoint)
                .WithRequestTimeout(this.options.ConnectionPolicy.RequestTimeout)
                .WithApplicationRegion(this.options.ConnectionPolicy.CurrentLocation)
                .WithThrottlingRetryOptions(TimeSpan.FromSeconds(this.options.ConnectionPolicy.RetryOptions.MaxRetryWaitTimeInSeconds), this.options.ConnectionPolicy.RetryOptions.MaxRetryAttemptsOnThrottledRequests)
                .BuildAndInitializeAsync(containersToInitialize)
                .ConfigureAwait(false);

When we are tracing the logs of cosmos db requests we are seeing this :
 "numberOfOpenTcpConnection": 10276
 "NumberOfClientsCreated": 6,
  "NumberOfActiveClients": 6,

**Comment by Pilchie** [comment]
When using Direct Mode, your machine will connect to all 4 replicas of each physical partition, from each of the clients.  The option of [MaxTcpConnectionsPerEndpoint](https://learn.microsoft.com/dotnet/api/microsoft.azure.cosmos.cosmosclientoptions.maxtcpconnectionsperendpoint?view=azure-dotnet) is for *each* backend node.

**Reply by Pilchie** [reply]
If you're trying to reduce the number of connections, why not use Gateway mode instead?

**Reply by kkattamudi-ms** [reply]
Despite having over 10,000 TCP connections, we are not achieving the desired throughput. We are currently in the analysis phase and are considering whether to increase the maximum number of TCP connections. We're trying to determine the optimal number of connections that can be configured without causing performance issues inside our Docker container or hitting system constraints.Can you please suggest what will be the idle number here  or a way to identify it?

Additionally, could you clarify the definition of an endpoint in this context? Does it refer to just one region or all the regions we have configured

**Reply by Pilchie** [reply]
What is your machine configuation?  Have you looked through the docs around this?

* https://learn.microsoft.com/azure/cosmos-db/nosql/best-practice-dotnet
* https://learn.microsoft.com/azure/cosmos-db/nosql/performance-tips-dotnet-sdk-v3
* https://learn.microsoft.com/azure/cosmos-db/nosql/tune-connection-configurations-net-sdk-v3

**Reply by kkattamudi-ms** [reply]
2 vcpus .
Attaching the recent log file for more insights .
[jarvislog.json](https://github.com/user-attachments/files/17235925/jarvislog.json)
This might help you to get more insights

**Reply by Pilchie** [reply]
What happens if you try the recommended settings in the above?

**Reply by kkattamudi-ms** [reply]
Currently we are running with 300+ machines but we are not able to scale to the desired throughput.

**Reply by kkattamudi-ms** [reply]
The incoming data is getting backlogged and we are seeing timeout exceptions sometimes from cosmosdb in our logs .

**Reply by Pilchie** [reply]
But have you applied the recommended settings?

**Reply by kkattamudi-ms** [reply]
Do you mean from the shared links  ?(https://learn.microsoft.com/azure/cosmos-db/nosql/tune-connection-configurations-net-sdk-v3) can you please share the specific settings if they are different?
